containers/src/Data/IntSet/Internal.hs

```diff
@@ -1157,8 +1158,156 @@ nequal _   _   = True
 --------------------------------------------------------------------}
 
 instance Ord IntSet where
-    compare s1 s2 = compare (toAscList s1) (toAscList s2)
-    -- tentative implementation. See if more efficient exists.
+    compare s1 s2 = orderingOf $ relateTop s1 s2
+
+-- | detailed outcome of lexicographic comparison of lists.
+-- w.r.t. Ordering, there are two extra cases,
+-- since (++) is not monotonic w.r.t. lex. order on lists
+-- (which is used by definition):
+-- consider comparison of  (Bin [0,3,4] [ 6] ) to  (Bin [0,3] [7] )
+-- where [0,3,4] > [0,3]  but [0,3,4,6] < [0,3,7].
+
+data Relation
+  = Less  -- ^ holds for [0,3,4] [0,3,5,1]
+  | Prefix -- ^ holds for [0,3,4] [0,3,4,5]
+  | Equals -- ^  holds for [0,3,4] [0,3,4]
+  | FlipPrefix -- ^ holds for [0,3,4] [0,3]
+  | Greater -- ^ holds for [0,3,4] [0,2,5]
+  deriving (Show, Eq)
+   
+orderingOf :: Relation -> Ordering
+{-# INLINE orderingOf #-}
+orderingOf r = case r of
+  Less -> LT
+  Prefix -> LT
+  Equals -> EQ
+  FlipPrefix -> GT
+  Greater -> GT
+
+-- The following gets complicated since integers are
+-- effectively handled (in the tree) by their binary representation:
+-- if a bit is zero, the left branch is taken.
+-- This also holds for the sign bit (the MSB),
+-- so negative numbers are in the right subtree:
+-- after    Bin p m l r = fromList [-1,0]
+-- we have  l = fromList [0], r = fromList [-1] .
+-- This can only happen at the very top, so handle this separetely,
+-- and avoid the check for the "mixed" case during recursion (function 'relate')
+-- We also avoid checking for Nil in 'relate', since it cannot appear below Bin.
+
+relateTop :: IntSet -> IntSet -> Relation
+{-# INLINE relateTop #-}
+relateTop Nil Nil = Equals
+relateTop Nil t2 = Prefix
+relateTop t1 Nil = FlipPrefix
+relateTop t1@(Bin p1 m1 l1 r1) t2@(Bin p2 m2 l2 r2)
+  | mixed t1 && mixed t2 = combine (relate r1 r2) (relate l1 l2)
+  | mixed t1 = combine_left (relate r1 t2)
+  | mixed t2 = combine_right (relate t1 r2)
+  | otherwise = relate t1 t2
+relateTop t1@(Bin p1 m1 l1 r1) t2@(Tip p2 bm2)
+  | mixed t1 = combine_left (relate r1 t2)
+  | otherwise = relate t1 t2
+relateTop t1@(Tip p1 bm1) t2@(Bin p2 m2 l2 r2)
+  | mixed t2 = combine_right (relate t1 r2)
+  | otherwise = relate t1 t2
+relateTop t1@(Tip _ _) t2@(Tip _ _) = relateTipTip t1 t2
+
+-- | precondition: each argument is non-Nil and non-mixed
+relate :: IntSet -> IntSet -> Relation
+relate t1@(Tip p1 bm1) t2@(Tip p2 bm2) = relateTipTip t1 t2
+relate t1@(Bin p1 m1 l1 r1) t2@(Bin p2 m2 l2 r2)
```

Comment:
In a bit more detail.... It seems to me that if `p1 < p2`, then `t1 ≤ t2`. Is that right? If so, then can we use the masks to determine (sometimes) that `t1 < t2`?

Reply:
I hope that this happens automatically by inlining `orderingOf $ relateTipTip ..`

Reply:
No, I mean pulling tricks in the `Bin/Bin` case. You can surely do the same thing you do with `Tip`: get lower and upper bounds for the trees, and if they don't overlap you have an answer.

Reply:
prefixes are tricky: assume wordSize is one (else multiply everthing by some large enough power of two), then fromList [2,3] has prefix 10 (binary) but fromList[3,4] has smaller prefix 0 (?)

Reply:
Okay, but can't you use your lower bound/upper bound calculation anyway? If the upper bound of one tree is less than the lower bound of the other, I think that's it.

Reply:
lower/upper-bound: yes this should be correct for any Tip/Bin combo:
```
  | succUpperbound t1 <= lowerbound t2 = Less
  | lowerbound t1 >= succUpperbound t2 = Greater
```
For now, this is inside the guards of some of the pattern matches but this could also be inverted (compare bounds first, and pattern match only if needed). This needs benchmarking (not today).

Reply:
I think that computing "lowerbound" inside a branch of a pattern match is better: the inliner then should be able to remove the pattern match that is in the implementation of "lowerbound"  (does it really?) 

I inserted the lower-upper test in the Bin/Bin case as helps to avoid recursion.